### PR TITLE
Pipeline

### PR DESCRIPTION
**Pull request by pomadchin**
TODO: 
- [x] Test at least in Zeppelin
- [x] Deprecate Spark ETL project
- [ ] Json Schema Validation
- [ ] User defined types
- [ ] Multiple inputs support (?)
- [ ] Tags (?)

* ? means - not sure that we really need it.

Chatta Demo example ingest (working): https://github.com/pomadchin/geotrellis-chatta-demo/blob/feature/pipeline/service/geotrellis/src/main/scala/geotrellis/chatta/ChattaPipelineIngest.scala

Cook book (json string parsing, interpretation & evaluation): 
```scala
// json => node => eval 
val json: String = ???

val node: Option[Node[Stream[(Int, TileLayerRDD[SpatialKey])]]] = 
  json.node[Stream[(Int, TileLayerRDD[SpatialKey])]]
val result: Option[Stream[(Int, TileLayerRDD[SpatialKey])]] = node.map(_.eval)

// json => eval
val result: Option[Stream[(Int, TileLayerRDD[SpatialKey])]] =
  json.eval[Stream[(Int, TileLayerRDD[SpatialKey])]]

// json => unsafeEval (without type checking)
val result: Option[Any] = json.unsafeEval

// Scala DSL example
import singleband.spatial._
val scheme = Left[LayoutScheme, LayoutDefinition](FloatingLayoutScheme(512))
val jsonRead = JsonRead("/", `type` = ReadTypes.SpatialHadoopType)
val jsonTileToLayout = TileToLayout(`type` = TransformTypes.SpatialTileToLayoutType)
val jsonReproject = Reproject("", scheme, `type` = TransformTypes.SpatialBufferedReprojectType)
val jsonPyramid = Pyramid(`type` = TransformTypes.SpatialPyramidType)

val jsonWrite = JsonWrite("write1", "/tmp", PipelineKeyIndexMethod("zorder"), scheme, `type` = WriteTypes.SpatialType)

val list = jsonRead ~ jsonTileToLayout ~ jsonReproject ~ jsonPyramid ~ jsonWrite

val typedAst =
  list
    .node[Stream[(Int, TileLayerRDD[SpatialKey])]]

typedAst.eval
```

### CODE REVIEW

**Comment by fosskers** [comment]
While I haven't reviewed the code, I have confirmed the behaviour through Zeppelin on an EMR cluster. The code I used:

```scala
import geotrellis.spark._
import geotrellis.spark.pipeline._
import geotrellis.spark.pipeline.ast._

val maskJson =
  """
    |[
    |  {
    |    "uri" : "s3://geotrellis-test/daunnc/LC81750172014163LGN00_LOW4.TIF",
    |    "time_tag" : "TIFFTAG_DATETIME",
    |    "time_format" : "yyyy:MM:dd HH:mm:ss",
    |    "type" : "singleband.spatial.read.s3"
    |  },
    |  {
    |    "resample_method" : "nearest-neighbor",
    |    "type" : "singleband.spatial.transform.tile-to-layout"
    |  },
    |  {
    |    "crs" : "EPSG:3857",
    |    "scheme" : {
    |      "crs" : "epsg:3857",
    |      "tileSize" : 256,
    |      "resolutionThreshold" : 0.1
    |    },
    |    "resample_method" : "nearest-neighbor",
    |    "type" : "singleband.spatial.transform.buffered-reproject"
    |  },
    |  {
    |    "end_zoom" : 0,
    |    "resample_method" : "nearest-neighbor",
    |    "type" : "singleband.spatial.transform.pyramid"
    |  },
    |  {
    |    "name" : "mask",
    |    "uri" : "s3://geotrellis-test/colingw/pipeline/",
    |    "key_index_method" : {
    |      "type" : "zorder"
    |    },
    |    "scheme" : {
    |      "crs" : "epsg:3857",
    |      "tileSize" : 256,
    |      "resolutionThreshold" : 0.1
    |    },
    |    "type" : "singleband.spatial.write"
    |  }
    |]
  """.stripMargin
  
  val list: Option[Node[Stream[(Int, TileLayerRDD[SpatialKey])]]] = maskJson.node
  
  list match {
      case None => println("Couldn't parse the JSON")
      case Some(node) => {
          node.eval.foreach { case (zoom, rdd) => println(s"ZOOM: ${zoom}"); println(s"COUNT: ${rdd.count}") }
      }
  }
  
  println("DONE")
```
This succeeds.

### Thoughts
First of all, using Spark through Zeppelin is great. It lets us script GeoTrellis functionality while submitting the actual work to an arbitrarily big EMR cluster in the background just by pressing a button.

The depths of the pipeline evaluation involve `Any` and some type casting. The example JSON above forces the use of `SpatialKey`, but if you fail to respect that in your type hint (say by putting `TemporalKey` instead), a "type error" will be given in the form of an exception. The exception itself is a bit cryptic:
```
java.lang.Exception: Cannot cast ErasedNode to scala.Stream[(Int, geotrellis.spark.TileLayerRDD[geotrellis.spark.TemporalKey])] since it cannot be cast to Stream[(Int, geotrellis.spark.TileLayerRDD[geotrellis.spark.SpatialKey])]
  at geotrellis.spark.pipeline.ast.untyped.ErasedNode$class.node(ErasedNode.scala:39)
  at geotrellis.spark.pipeline.ast.untyped.ErasedNodeComposition.node(ErasedNode.scala:81)
  at geotrellis.spark.pipeline.ast.untyped.Implicits$withStringExtensions$$anonfun$node$1.apply(Implicits.scala:26)
  at geotrellis.spark.pipeline.ast.untyped.Implicits$withStringExtensions$$anonfun$node$1.apply(Implicits.scala:26)
  at scala.Option.map(Option.scala:146)
...
```
It would nice if there were a pure variant of `.eval` that returned an `Either`, so that we're up-front about the fact that `.eval` can fail.

**Comment by pomadchin** [comment]
Empty folder bug solution: the problem is in a trailing slash, replace `"uri" : "s3://geotrellis-test/colingw/pipeline/"` with `"uri" : "s3://geotrellis-test/colingw/pipeline"`.

**Comment by pomadchin** [comment]
An answer to the confusing size of outputs (Multiband ingest).
P.S. All sizes are matching to the same ingest using `Ingest` object,

```
import geotrellis.spark._
import geotrellis.spark.pipeline._
import geotrellis.spark.pipeline.ast._

implicit val _sc = sc

val maskJson =
  """
    |[
    |  {
    |    "uri" : "s3://geotrellis-test/daunnc/LC81750172014163LGN00_LOW4.TIF",
    |    "time_tag" : "TIFFTAG_DATETIME",
    |    "time_format" : "yyyy:MM:dd HH:mm:ss",
    |    "type" : "multiband.spatial.read.s3"
    |  },
    |  {
    |    "resample_method" : "nearest-neighbor",
    |    "type" : "multiband.spatial.transform.tile-to-layout"
    |  },
    |  {
    |    "crs" : "EPSG:3857",
    |    "scheme" : {
    |      "crs" : "epsg:3857",
    |      "tileSize" : 256,
    |      "resolutionThreshold" : 0.1
    |    },
    |    "resample_method" : "nearest-neighbor",
    |    "type" : "multiband.spatial.transform.buffered-reproject"
    |  },
    |  {
    |    "end_zoom" : 0,
    |    "resample_method" : "nearest-neighbor",
    |    "type" : "multiband.spatial.transform.pyramid"
    |  },
    |  {
    |    "name" : "LC81750172014163LGN00_LOW4_MB",
    |    "uri" : "s3://geotrellis-test/daunnc/pipeline",
    |    "key_index_method" : {
    |      "type" : "zorder"
    |    },
    |    "scheme" : {
    |      "crs" : "epsg:3857",
    |      "tileSize" : 256,
    |      "resolutionThreshold" : 0.1
    |    },
    |    "type" : "multiband.spatial.write"
    |  }
    |]
  """.stripMargin
  
  val list: Option[Node[Stream[(Int, MultibandTileLayerRDD[SpatialKey])]]] = maskJson.node
  
  list match {
      case None => println("Couldn't parse the JSON")
      case Some(node) => {
          node.eval.foreach { case (zoom, rdd) => println(s"ZOOM: ${zoom}") }
      }
  }
  
  val reader = LayerReader("s3://geotrellis-test/daunnc/pipeline")

  val layerId = LayerId("LC81750172014163LGN00_LOW4_MB", 7)
  val rdd = reader.read[SpatialKey, MultibandTile, TileLayerMetadata[SpatialKey]](layerId)
  val raster = rdd.stitch
  val tiff = GeoTiff(raster, rdd.metadata.crs)

  tiff.write(new Path("s3://geotrellis-test/daunnc/pipeline/test-zoom-mb.tif"))
```

File: spark-pipeline/src/main/scala/geotrellis/spark/pipeline/json/transform/Transform.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package geotrellis.spark.pipeline.json.transform
+
+import geotrellis.raster.resample.{NearestNeighbor, PointResampleMethod}
+import geotrellis.raster.CellType
+import geotrellis.spark.pipeline.json._
+import geotrellis.spark.tiling.{LayoutDefinition, LayoutScheme}
+import io.circe.generic.extras.ConfiguredJsonCodec
+
+trait Transform extends PipelineExpr
```
Comment (moradology) [inline spark-pipeline/src/main/scala/geotrellis/spark/pipeline/json/transform/Transform.scala]: We already discussed this a bit. This is a reminder to think about removing these classes and just using circe encoder/decoder logic to manage serialization/deserialization

File: spark-pipeline/src/main/scala/geotrellis/spark/pipeline/ast/Node.scala [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package geotrellis.spark.pipeline.ast
+
+import geotrellis.spark.pipeline.json.PipelineExpr
+
+import io.circe.Json
+import org.apache.spark.SparkContext
+
+trait Node[T] extends Serializable {
+  def eval(implicit sc: SparkContext): T
```
[the remark this answers is not in the document]
Comment (pomadchin) [reply spark-pipeline/src/main/scala/geotrellis/spark/pipeline/ast/Node.scala]: Yeah, probably there should be two eval functions (`eval` (for collections) and `evalSpark(implicit sc: SparkContext)`)?

File: spark-pipeline/src/main/scala/geotrellis/spark/pipeline/ast/Node.scala [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  def arg: PipelineExpr
+  def asJson: List[Json]
```
[the remark this answers is not in the document]
Comment (pomadchin) [reply spark-pipeline/src/main/scala/geotrellis/spark/pipeline/ast/Node.scala]: That's a good question. 
P.S. `T` here doesn't represent a `JSON` type.

**Comment by metasim** [comment]
> Overall I'm not hugely convinced that writing JSON like this is easier than writing Scala, but if other tools were written that could produce this JSON, then there's probably value in that. That would open the door for libraries in other languages to take advantage of GeoTrellis ingest capabilities.

Sorry for the unsolicited comment, but this would be my suspicion too, given that you could be using case classes that are compiled on-the-fly (something I've done a number of times, and could help with the minor trickery involved), and end up with something that is just as readable if not more, is type safe, and doesn't require a separate internal object representation. Could be extended with DSL-like features via extension methods, etc. to make it even more rich than JSON will allow.

**Comment by metasim** [comment]
PPS: With Zeppelin integration you can create a custom interpreter type where a configuration DSL/model is automatically imported and give you some of the code completion features there too (although not as good as an IDE).

**Comment by pomadchin** [comment]
@metasim What means `doesn't require a separate internal object representation` and `compiled on-the-fly`? Is it possible to expand it a bit? 

Confused a bit about comments focused on JSON, as Pipeline is a replacement to ETL package, and we have to deal somehow with CLI / outer world `strings` representation. There is the goal to keep typesafety as much as it's possible even taking into account that the real world is not typesafe at all, and to make it possible to use CLI (in fact here we have a ready interface for CLI) without understanding Scala at all. But if user would decide to use Scala why not, there is Scala Pipeline DSL, core API, etc?

File: spark-pipeline/src/main/scala/geotrellis/spark/pipeline/ast/RealWorld.scala [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package geotrellis.spark.pipeline.ast
+
+import geotrellis.spark.pipeline.json.PipelineExpr
+
+import io.circe.Json
+import org.apache.spark.SparkContext
+
+sealed trait RealWorld
```
[the remark this answers is not in the document]
Comment (pomadchin) [reply spark-pipeline/src/main/scala/geotrellis/spark/pipeline/ast/RealWorld.scala]: I just wanted user to understand, that it doesn't meant that you can pass `null` or `Unit` inside it, it means IO operation, to read smth from the entire world, in fact a hack to make type system clean; 

I was inspired by a Haskell `RealWorld` type from the `State monad`: http://hackage.haskell.org/package/base-4.10.0.0/docs/Control-Monad-ST.html#t:RealWorld

**Comment by metasim** [comment]
@pomadchin Here's the kind of thing I'm talking about:

https://github.com/metasim/scala-as-config/blob/master/src/main/scala/astraea/example/ScalaAsConfig.scala

**Comment by pomadchin** [comment]
@metasim nice, mb makes sense for user defined functions in some way

**Comment by metasim** [comment]
@pomadchin Was it clear how one could have external text files, just like JSON, loaded, type-checked, and compiled at run time? IOW, I think the benefits are much more extensive than providing hooks for user-defined functions.

**Comment by echeipesh** [comment]
I think this is good to be merged and leaned on. Next steps here would be to do some user-testing and create relevant documentation to replace `spark-etl` package.

**Pull request merged**